### PR TITLE
owner(cdc): add a new metadata interface

### LINKED ISSUE #9461: support more metadata storage like RDS
### Is your feature request related to a problem?

Currently TiCDC metadata is stored in ETCD, which is not very scalable. And if changefeed progress is stored in a vector, ETCD doesn't have very strong transaction support.

### Describe the feature you'd like

Design a new metadata storage interface to support multiple backends.

### Describe alternatives you've considered

_No response_

### Teachability, Documentation, Adoption, Migration Strategy

_No response_

### PR DESCRIPTION
**Pull request by hicqu**
### What problem does this PR solve?

Issue Number: close #9461 

### What is changed and how it works?

Add a new metadata interface.

#### Tests 

 - No code

### Release note 

```release-note
None
```

### CODE REVIEW

File: cdc/metadata/metadata.go [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+	CreateChangefeed(cf *ChangefeedInfo, up *UpstreamInfo) error
+
+	// UpdateChangefeed updates changefeed metadata, must be called on a stopped one.
+	UpdateChangefeed(cf *ChangefeedInfo) error
+
+	// Stop a changefeed.
+	SetChangefeedStopped(cf *ChangefeedInfo) error
+
+	// Fetch the latest capture list in the TiCDC cluster.
+	RefreshCaptures() <-chan []*CaptureInfo
```
Comment (CharlesCheung96) [inline cdc/metadata/metadata.go:126]: This method should be placed alongside the heartbeat method.

File: cdc/metadata/metadata.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+// All intrefaces are thread-safe and shares one same Context.
+type ControllerObservation interface {
+	// CreateChangefeed creates a changefeed.
+	CreateChangefeed(cf *ChangefeedInfo, up *UpstreamInfo) error
+
+	// UpdateChangefeed updates changefeed metadata, must be called on a stopped one.
+	UpdateChangefeed(cf *ChangefeedInfo) error
+
+	// Stop a changefeed.
+	SetChangefeedStopped(cf *ChangefeedInfo) error
```
Comment (CharlesCheung96) [inline cdc/metadata/metadata.go:123]: Duplicate with UpdateChangefeed? In addition, we should restrict the cahngefeed state only at the api level.

File: cdc/metadata/metadata.go [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+	Advance(cfs []*ChangefeedInfo, progresses []*ChangefeedProgress) error
+
+	// Fetch the latest alive changefeed owner list from metadata.
+	RefreshAliveOwners() <-chan []*ChangefeedInfo
+}
+
+// ControllerObservation is for observing and updating meta by Controller.
+//
+// All intrefaces are thread-safe and shares one same Context.
+type ControllerObservation interface {
```
Comment (CharlesCheung96) [inline cdc/metadata/metadata.go:115]: How to remove changefeed?

File: cdc/metadata/metadata.go [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+// OwnerObservation is for observing and updating running status of a changefeed.
+//
+// All intrefaces are thread-safe and shares one same Context.
+//
+// On a capture all owners shares one same OwnerObservation, because heartbeats
+// to metadata storage is based on CAPTURE instead of CHANGEFEED.
+type OwnerObservation interface {
+	// Advance advances a changefeed progress.
+	Advance(cf *ChangefeedInfo, progress *ChangefeedProgress) error
```
Comment (CharlesCheung96) [inline cdc/metadata/metadata.go:150]: Is this duplicate with `CaptureObservation .Advance`?

File: cdc/metadata/metadata.go [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+	// set the changefeed to state finished.
+	SetChangefeedFinished(cf *ChangefeedInfo) error
```
Comment (CharlesCheung96) [inline cdc/metadata/metadata.go:153]: Perhaps `updateChangefeedState` is sufficient to handle these scenarios?

File: cdc/metadata/metadata.go [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+	// Fetch the latest alive changefeed owner list from metadata.
+	RefreshAliveOwners() <-chan []*ChangefeedInfo
+}
+
+// ControllerObservation is for observing and updating meta by Controller.
+//
+// All intrefaces are thread-safe and shares one same Context.
+type ControllerObservation interface {
+	// CreateChangefeed creates a changefeed.
+	CreateChangefeed(cf *ChangefeedInfo, up *UpstreamInfo) error
```
Comment (hicqu) [inline cdc/metadata/metadata.go:117]: NOTE: operations on changefeeds won't be batched. Maybe we need to add methods like `PauseChangefeeds(...)`.

File: cdc/metadata/metadata.go [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+	// TakeControl blocks until becomes controller or gets canceled.
+	TakeControl() ControllerObservation
+
+	// Advance is like OwnerObservation.Advance but handles many changefeeds.
+	//
+	// It should returns an InvalidOwner error if any changefeed ownership is revoked.
+	Advance(cfs []*ChangefeedInfo, progresses []*ChangefeedProgress) error
+
+	// Fetch the latest alive changefeed owner list from metadata.
+	RefreshAliveOwners() <-chan []*ChangefeedInfo
```
Comment (hicqu) [inline cdc/metadata/metadata.go:109]: Add more interfaces to query changefeeds or upstreams.

Comment (CharlesCheung96) [inline cdc/metadata/metadata.go:109]: How do I get the corresponding changefeed progress of the owner?

File: cdc/metadata/metadata.go [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+// 5. Controller can call ControllerObservation.LaunchOwner to re-scheduling it.
+//
+// So there is a state for changefeed owner on metadata storage:
+//   stopped -> launched -> stopping -> stoped.
+// ---------------------------------------------------------------- //
+
+// ---------- About changefeed worker captures scheduling ---------- //
+// 1. ControllerObservation.AttachWorkers attaches some captures to a changefeed;
+// 2. Owner calls OwnerObservation.RefreshWorkers to know how to launch or stop processors;
+// 3. Capture receives processor launch/stop messages with P2P mechanism;
```
Comment (CharlesCheung96) [inline cdc/metadata/metadata.go:69]: With P2P, we can't handle network isolation.

File: cdc/metadata/metadata.go (hunk 7 above)
Comment (sdojjy) [inline cdc/metadata/metadata.go:109]: some APIs need resolvedTs

File: cdc/metadata/metadata.go [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+	RefreshProcessors() <-chan []ScheduledProcessor
+
+	// When a processor exits, inform the metadata storage.
+	PostProcessorStopped(cf *ChangefeedInfo) error
+}
+
+// ControllerObservation is for observing and updating meta by Controller.
+//
+// All intrefaces are thread-safe and shares one same Context.
+type ControllerObservation interface {
```
Comment (CharlesCheung96) [inline cdc/metadata/metadata.go:162]: Need GC scheduler info

**Review by sdojjy — APPROVED** [review]

File: cdc/metadata/metadata.go [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+	// Notes:
+	//   * the target capture can fetch the event by `RefreshOwners`.
+	//   * select `done` to wait the old owner in stopping to be resolved.
+	SetOwner(cf *ChangefeedInfo, target *CaptureInfo) (done <-chan struct{}, _ error)
+
+	// Schedule some captures as workers to a given changefeed.
+	// Notes:
+	//   * target captures can fetch the event by `RefreshProcessors`.
+	//   * select `done` to wait all processors in stopping to be resolved.
+	SetProcessors(cf *ChangefeedInfo, workers []*CaptureInfo) (done <-chan struct{}, _ error)
```
Comment (CharlesCheung96) [inline cdc/metadata/metadata.go:182]: The caller needs to pass in whether processor is in the removing state.

Comment (CharlesCheung96) [inline cdc/metadata/metadata.go:182]: Why does the Controller need to monitor the state of the processor?

Comment (hicqu) [reply cdc/metadata/metadata.go]: `done` is for being uniform with `SetOwner`. We can just print a log after processors in stopping get resolved.

File: cdc/metadata/metadata.go (hunk 9 above)
Comment (hicqu) [reply cdc/metadata/metadata.go:162]: GC should be handled by controller internally.

**Review by CharlesCheung96 — APPROVED** [review]

File: cdc/metadata/metadata.go [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+// All intrefaces are thread-safe and shares one same Context.
+type ControllerObservation interface {
+	// CreateChangefeed creates a changefeed.
+	CreateChangefeed(cf *ChangefeedInfo, up *UpstreamInfo) error
+
+	// RemoveChangefeed removes a changefeed, will auto stop owner and processors.
+	RemoveChangefeed(cf *ChangefeedInfo) error
+
+	// Fetch the latest capture list in the TiCDC cluster.
+	RefreshCaptures() <-chan []*CaptureInfo
```
Comment (CharlesCheung96) [inline cdc/metadata/metadata.go:183]: If CaptureInfo  returned in GetChangefeedSchedule, RefreshCaptures may not be required?

File: cdc/metadata/metadata.go [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+	ChangefeedInfo() *ChangefeedInfo
+
+	// PauseChangefeed pauses a changefeed.
+	PauseChangefeed() error
+
+	// ResumeChangefeed resumes a changefeed.
+	ResumeChangefeed() error
+
+	// UpdateChangefeed updates changefeed metadata, must be called on a paused one.
+	UpdateChangefeed(*ChangefeedInfo) error
```
Comment (CharlesCheung96) [inline cdc/metadata/metadata.go:216]:
```suggestion
	UpdateChangefeedInfo(*ChangefeedInfo) error
```

**Pull request merged**